contracts/exchange/exchange_state.cpp

```diff
@@ -0,0 +1,87 @@
+#include <exchange/exchange_state.hpp>
+
+namespace eosio {
+   extended_asset exchange_state::convert_to_exchange( connector& c, extended_asset in ) {
+
+      real_type R(supply.amount);
+      real_type C(c.balance.amount+in.amount);
+      real_type F(c.weight/1000.0);
+      real_type T(in.amount);
+      real_type ONE(1.0);
+
+      real_type E = -R * (ONE - std::pow( ONE + T / C, F) );
```

Comment:
Hi Dan,
isn't std::pow dangerous in the context of distributed consensus?
We've generally avoided using floating point maths in graphene for that reason, and std::pow seems particularly dangerous to me, because it's always just an approximation algorithm.

Reply:
floating point math is not inherently bad.  The concern is that there are enough differences in hardware implementations that it is a vector to non-determinism due to precision, rounding, and most importantly the bit-value of all the invalid or control values like NaN.  Non-determinism is very bad.  

Which is simply to say, if you can guarantee that its not an avenue for non-determinism then we can use it despite its nature as an approximation.  Or rather, the approximation is consistently inaccurate 🙈 🙊 🙉 

We have achieved this by using a software float library which essentially makes its _deterministic_ handling of rounding, precision, and special values standard behavior in WebAssembly contracts.  

We have changes coming in soon that will effectively transpile the WebAssembly native floating point instructions to our available soft_float intrinsics on the fly.   Meaning that most contract devs wont have to do anything explicit to take advantage of the feature.

Reply:
SoftFloat from Berkeley is a great deterministic implementation of Floating-Point Arithmetic operations